lib/src/HttpServer.cc

```diff
@@ -489,49 +461,107 @@ void HttpServer::onRequests(
+                                auto resp = requestParser->getFirstResponse();
+                                if (resp.first)
                                 {
-                                    auto resp =
-                                        requestParser->getFirstResponse();
-                                    if (resp.first)
-                                    {
-                                        requestParser->popFirstRequest();
-                                        resps.push_back(std::move(resp));
-                                    }
-                                    else
-                                        break;
+                                    requestParser->popFirstRequest();
+                                    resps.push_back(std::move(resp));
                                 }
-                                sendResponses(conn,
-                                              resps,
-                                              requestParser->getBuffer());
-                            }
-                            else
-                            {
-                                // some earlier requests are waiting for
-                                // responses;
-                                requestParser->pushResponseToPipelining(
-                                    req, newResp, isHeadMethod);
+                                else
+                                    break;
                             }
+                            sendResponses(conn,
+                                          resps,
+                                          requestParser->getBuffer());
                         }
-                    });
-                }
-            });
+                        else
+                        {
+                            // some earlier requests are waiting for
+                            // responses;
+                            requestParser->pushResponseToPipelining(
+                                req, newResp, isHeadMethod);
+                        }
+                    }
+                });
+            }
+        };
+
+        const bool enableDecompression = app().isCompressedRequestEnabled();
+        bool sendForProcessing = true;
+        if (enableDecompression)
+        {
+            auto contentEncoding = req->getHeaderBy("content-encoding");
+            auto body = req->getBody();
+            if (contentEncoding == "")
+            {
+                // NoOP
+            }
+            else if (contentEncoding == "gzip")
+            {
+                req->setBody(utils::gzipDecompress(body.data(), body.size()));
+                req->removeHeader("content-encoding");
+            }
+#ifdef USE_BROTLI
+            else if (contentEncoding == "br")
+            {
+                req->setBody(utils::brotliDecompress(body.data(), body.size()));
+                req->removeHeader("content-encoding");
+            }
+#endif
```

Comment:
@an-tao Do think we need a new function to directly decompress into a `CacheFile`?-Otherwise decompressing can causes a very short lived memory spike.

Reply:
yes, we could write decompressed data into cache files when the `body` is already in a cache file.

Reply:
I made it so we store the decompressed result in CacheFile if the resulting size is larger than max body memory size. I also fixed the memory spike issue by stopping decompress right when we find the result is too large.